### ISSUE TITLE
Would be nice to have session.fetch(...).asPromise()

### ISSUE THREAD

**Issue by cypherix93** [artifacthealth/hydrate-mongodb#7]
Hey there, I've been using this library for a bit now and I love the way every query can be promisified so I can use `await` on them. However, the `fetch` function on `Session` does not have an `asPromise` function. I would like to have something like the following:

```
var stuff = new Stuff();
session.save(stuff);
session.flush();

// I would like to have the new Stuff document from the DB at this point.
var dbStuff = await session.fetch(stuff).asPromise();

return dbStuff; // as JSON to front end
```

Is it possible to do something like this?

Thanks!

**Comment by meirgottlieb** [comment]
Yeah, that should be OK. I'll work it into the next release.

**Comment by meirgottlieb** [comment]
Sorry, I was too quick to respond. Because of the way operations on the session are queued up this is actually I bigger change than I thought. I'm going to have to think about this one. I'll leave the issue option. In the meantime, you can chain `fetch` with a query and get a promise. For example:

```
var fetchedStuff = await session.find(Stuff, id).fetch("someReference").asPromise();
```

**Comment by cypherix93** [comment]
Thanks for the prompt responses. The issue I am facing is getting a hold of the `_id` after a `session.save(...)` call. I tried looking at the same object reference that was saved, but the `_id` or `id` field wasn't populated. That's why I needed to do a fetch. Any tips on how to accomplish that?

I can't use `session.find(Stuff, id).fetch(...)` because I don't have the id.

**Comment by meirgottlieb** [comment]
The call to `save` is async. Although the callback is not required, the `id` is assigned during the save operation. If you put your code in the callback to `save`, the `id` will be available.

**Comment by cypherix93** [comment]
Would it then be desirable to have an `asPromise()` for `save` as well? Or maybe if the second argument is not passed, it returns a `Promise` by default? I'll try to put together a PR if I can manage the time.

**Comment by meirgottlieb** [comment]
Yeah, whatever I do for `fetch` I'll do for the rest of the methods on `Session` as well. What I'm leaning toward is returning a Promise whenever a callback is not provided.

**Comment by cypherix93** [comment]
That sounds great! Thanks.

**Comment by cypherix93** [comment]
Maybe a solution would be to update the tracked references after flush is called? I don't know how exactly the tracking is implemented in this library, but coming from a C# background, Entity Framework injects the database-generated fields into the tracked model upon calling `SaveChanges()`.

I think an elegant solution would be to keep the current implementation that you have, and implement something similar to what Entity Framework does.

Example for clarity:

```
var stuff = new Stuff();
//... fill in Stuff fields

session.save(stuff); // goes into the queue for later execution

session.flush(); // executes the save

// at this point the original reference to stuff should have the _id field
console.log(stuff._id) // => "587524fdedab1f0014ff0cb2"
```

What do you think?

**Comment by meirgottlieb** [comment]
That's pretty much how it works now. The issue is each of those steps is async and the current way the Session handles errors doesn't exactly mesh with returning Promises, but I have a few ideas. I'll update this thread when I have more.

**Comment by edcarroll** [comment]
Sorry to press you, I'm very keen to be able to use this functionality - have you had any further progress with it? :smile:

**Comment by meirgottlieb** [comment]
Doing this will require a breaking change to the API. I haven't put it in place yet.
